scripts/igmp_query_injector.py

```diff
@@ -0,0 +1,445 @@
+#!/usr/bin/env python
+import Queue
+import argparse
+import os
+import socket
+import struct
+import subprocess
+import threading
+from abc import ABCMeta, abstractmethod
+import logging
+import signal
+import resource
+import binascii
+from xcp import logger as log
+import time
+import sys
+import re
+from xen.lowlevel.xs import xs
+
+
+class Cleanup(object):
+    """Ensure xenstore connection closed when exit
+    """
+    CATCH_SIGNALS = [signal.SIGTERM, signal.SIGINT]
+
+    def __enter__(self):
+        self._set_signal_handler(self.CATCH_SIGNALS, self._close_xs_connection)
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        XSUtil.close()
+
+    def _set_signal_handler(self, signals, f):
+        for s in signals:
+            signal.signal(s, f)
+
+    def _close_xs_connection(self, signum, frame):
+        log.critical('Catch signal %d, close xenstore connection' % signum)
+        XSUtil.close()
+        sys.exit(ERRNO.SIGNAL_CATCH)
+
+
+class ERRNO(object):
+    SUCC = 0
+    SIGNAL_CATCH = 1
+
+
+class XSUtil(object):
```

Comment:
I don't think this class adds much value.   I think it would be simpler to call the Xenstore library directly when you need it, and pass around the connection as needed.